core/src/main/scala/kafka/coordinator/group/GroupCoordinator.scala

```diff
@@ -1104,14 +1107,16 @@ class GroupCoordinator(val brokerId: Int,
             group.maybeInvokeJoinCallback(member, JoinGroupResult(member.memberId, Errors.NOT_COORDINATOR))
           }
 
-          joinPurgatory.checkAndComplete(GroupKey(group.groupId))
+          rebalancePurgatory.checkAndComplete(GroupJoinKey(group.groupId))
```

Comment:
Not related to this patch, but I'm wondering if this is sufficient to expire the DelayedJoin. The completion condition is `hasAllMembersJoined` which may well be false when the group is unloaded.

On a related note, we should probably consider adding generationId to `DelayedJoin` as well, but we can consider that separately.

Reply:
> Not related to this patch, but I'm wondering if this is sufficient to expire the DelayedJoin. The completion condition is hasAllMembersJoined which may well be false when the group is unloaded.

Good point. I will check that separately. 

> On a related note, we should probably consider adding generationId to DelayedJoin as well, but we can consider that separately.

Good idea. I have filed a JIRA for this: https://issues.apache.org/jira/browse/KAFKA-12961.